ProcessLib/HydroMechanics/HydroMechanicsFEM.h

```diff
@@ -522,34 +486,52 @@ class HydroMechanicsLocalAssembler : public LocalAssemblerInterface
         return getIntPtEpsilon(cache, 5);
     }
 
-    std::vector<double> const& getIntPtDarcyVelocityX(
-        const double /*t*/,
-        GlobalVector const& /*current_solution*/,
-        NumLib::LocalToGlobalIndexMap const& /*dof_table*/,
-        std::vector<double>& /*cache*/) const override
+    std::vector<double> const& getIntPtDarcyVelocity(
+        const double t,
+        GlobalVector const& current_solution,
+        NumLib::LocalToGlobalIndexMap const& dof_table,
+        std::vector<double>& cache) const override
     {
-        assert(!_darcy_velocities.empty());
-        return _darcy_velocities[0];
-    }
+        auto const num_intpts = _ip_data.size();
 
-    std::vector<double> const& getIntPtDarcyVelocityY(
-        const double /*t*/,
-        GlobalVector const& /*current_solution*/,
-        NumLib::LocalToGlobalIndexMap const& /*dof_table*/,
-        std::vector<double>& /*cache*/) const override
-    {
-        assert(_darcy_velocities.size() > 1);
-        return _darcy_velocities[1];
-    }
+        auto const indices = NumLib::getIndices(_element.getID(), dof_table);
+        assert(!indices.empty());
+        auto const local_x = current_solution.get(indices);
 
-    std::vector<double> const& getIntPtDarcyVelocityZ(
-        const double /*t*/,
-        GlobalVector const& /*current_solution*/,
-        NumLib::LocalToGlobalIndexMap const& /*dof_table*/,
-        std::vector<double>& /*cache*/) const override
-    {
-        assert(_darcy_velocities.size() > 2);
-        return _darcy_velocities[2];
+        cache.clear();
+        auto cache_vec = MathLib::createZeroedMatrix<Eigen::Matrix<
```

Comment:
Minor: I'd prefer `cache_mat` since it's a matrix. Probably I introduced it wrongly.

Reply:
This is HM. @wenqing just told me that it's going to be split to a separate PR.

Reply:
@chleh Found the problem in the HM part. The bug is not caused by the order of shape function but by setting a wrong number of the components in  ``makeExtrapolator`` for velocity. With the bug fixing, the HM part is kept in this PR, and I am not going to bloat the source code with hundreds line change any more :smiley: . 

BTW, cache_vec is changed to cache_matrix as well.